eras/shelley/impl/src/Cardano/Ledger/Shelley/Rules/Tick.hs

```diff
@@ -167,6 +175,63 @@ validatingTickTransition nes slot = do
 
   pure $ nes' {nesEs = es''}
 
+-- | This is a limited version of 'validatingTickTransition' which is only suitable
+-- for the future ledger view.
+validatingTickTransitionFORECAST ::
+  forall tick era.
+  ( State (tick era) ~ NewEpochState era,
+    BaseM (tick era) ~ ShelleyBase,
+    State (EraRule "UPEC" era) ~ UpecState era,
+    Signal (EraRule "UPEC" era) ~ (),
+    Environment (EraRule "UPEC" era) ~ EpochState era,
+    Embed (EraRule "UPEC" era) (tick era),
+    STS (tick era)
+  ) =>
+  NewEpochState era ->
+  SlotNo ->
+  TransitionRule (tick era)
+validatingTickTransitionFORECAST nes slot = do
+  -- This whole function is a specialization of an inlined 'NEWEPOCH'.
+  --
+  -- The ledger view, 'LedgerView', is built entirely from the 'nesPd' and 'esPp' and
+  -- 'dsGenDelegs', so the correctness of 'validatingTickTransitionFORECAST' only
+  -- depends on getting these three fields correct.
+
+  epoch <- liftSTS $ do
+    ei <- asks epochInfoPure
+    epochInfoEpoch ei slot
+
+  let es = nesEs nes
+      ss = esSnapshots es
+
+  -- the relevant 'NEWEPOCH' logic
+  let pd' = ssStakeMarkPoolDistr ss
```

Comment:
I know that `pd'` is lazy, but for clarity that it is not used only in the `else` clause, maybe move it there as well